paper_framework_sources/p2079_system_execution_context.md

```diff
@@ -178,14 +181,163 @@ public:
+The downsides are that it is hard to ensure that the right context is set early enough in the process and that, like link-time replacement, it requires type erasure.
+
+The other question is to what extent we need to specify this.
+We could simply say that implementations should allow customization and leave it up to QOI.
+We already know that full platform customisations are possible.
+This approach would delegate the decision of how to allow Intel to replace the platform context with TBB up to the platform implementor.
+It would rely on an agreement between the system vendor and the runtime vendor.
+
+The authors do not have a recommendation, only a wish to see customisation available.
+We should decide how best to achieve it within the standard.
+Assuming we delegate customisation to the platform implementor, what wording would be appropriate for the specification, if any?
+
+## Need for the system_context class
+Our goal is to expose a global shared context to avoid oversubscription of threads in the system and to efficiently share a system thread pool.
+Underneath the `system_context` there is a singleton of some sort, potentially owned by the OS.
+
+The question is how we expose the singleton.
+We have a few obvious options:
+
+ - Explicit context objects, as we've described in R2 and R3 of this paper, where a `system_context` is constructed as any other context might be, and refers to a singleton underneath.
+ - A global `get_system_context()` function that obtains a `system_context` object, or a reference to one, representing the singleton explicitly.
+ - A global `get_system_scheduler()` function that obtains a scheduler from some singleton system context, but does not explicitly expose the context.
+
+The `get_system_context()` function returning by value adds little, it's equivalent to direct construction.
+`get_system_context()` returning by reference and `get_system_scheduler()` have a different lifetime semantic from directly constructed `system_context`.
+
+The main reason for having an explicit by-value context is that we can reason about lifetimes.
+If we only have schedulers, from `get_system_context().get_scheduler()` or from `get_system_scheduler()` then we have to think about how they affect the context lifetime.
+We might want to reference count the context, to ensure it outlives the schedulers, but this adds cost to each scheduler use, and to any downstream sender produced from the scheduler as well that is logically dependent on the scheduler.
+We could alternatively not reference count and assume the context outlives everything in the system, but that leads quickly to shutdown order questions and potential surprises.
+
+By making the context explicit we require users to drain their work before they drain the context.
+In debug builds, at least, we can also add reference counting so that descruction of the context before work completes reports a clear error to ensure that people clean up.
```

Comment:
I am not sure we should suggest different approaches in Debug and Release if they imply different ABI. From the previous experience it's real burden debug build with Microsoft STL is incompatible with Release build with Microsoft STL that forces an user to have yet another set of binaries.

I don't think it is fundamental change though. Let's keep it for now and let's see how discussion in SG1 goes.